### PR TITLE
Objectify FanGraphs 2.0

### PR DESCRIPTION
**Pull request by TheCleric**
v2.0 of #114 with a lot of feedback from @schorrm and @bdilday integrated in.

**NOTE:** At @schorrm 's request (to keep this PR smaller) I've removed the unit testing. I do have those coded and working. They will come in #121 after this one is merged.

This PR converts all the FanGraphs calls under a unified methodology. This accomplishes a few things:

Splits out the HTML parsing so that it can be extended for other data sources such as BRef now or Seamheads later.
Since all the FanGraphs data revolves around the /leaders.aspx page with different params, this collapses them into one root function.
Specifies a lot of the options via enums so that type checking is a bit easier. Included in this are complete enums of the supported FanGraphs columns for Batting, Fielding, and Pitching.

### CODE REVIEW

**Comment by schorrm** [comment]
No, I didn't mean remove the unit testing, I meant leave it as is (and intentionally fail it, just so the file count wouldn't be so high)

**Comment by TheCleric** [comment]
> No, I didn't mean remove the unit testing, I meant leave it as is (and intentionally fail it, just so the file count wouldn't be so high)

Sorry. I'm allergic to broken unit tests. 🤣 

I'll revert them shortly.

I'm not sure if the PRs will be mergable with broken tests, so let me know if I need disable the tests check as well and re-enable them in the next PR.

**Comment by schorrm** [comment]
It better be mergable with broken unit tests. I want the diagnostics, I don't want the CI telling _me_ what to do. M

**Comment by TheCleric** [comment]
> It better be mergable with broken unit tests. I want the diagnostics, I don't want the CI telling _me_ what to do. M

I do. Most developers I know can't be trusted. 🤣

File: pybaseball/batting_leaders.py
Comment (schorrm) [inline pybaseball/batting_leaders.py:33]: okay whoaaa what is this?

Comment (schorrm) [inline pybaseball/batting_leaders.py:33]: if this is the same, why isn't `batting_stats` the same as `fg_batting_data`? why do we need separate hooks for this?

File: pybaseball/datasources/fangraphs.py
Comment (schorrm) [inline pybaseball/datasources/fangraphs.py:76]: aka `str(int(split_seasons))` or for that matter `f'{split_seasons:d}'`, but the whole thing seems unnecessary. Why not just leave this as an int?

Comment (schorrm) [inline pybaseball/datasources/fangraphs.py:76]: aka `str(int(split_seasons))` or for that matter `f'{split_seasons:d}'`, but the whole thing seems unnecessary. Why not just leave this as an int?

File: pybaseball/datasources/html_table_processor.py
Comment (schorrm) [inline pybaseball/datasources/html_table_processor.py:68]: Can this be anything other than 200 and still pass?

File: pybaseball/batting_leaders.py
Comment (schorrm) [inline pybaseball/batting_leaders.py:33]: what is this?

Comment (schorrm) [inline pybaseball/batting_leaders.py:33]: if this is the same, why isn't `batting_stats` the same as `fg_batting_data`? why do we need separate hooks for this?

File: pybaseball/team_batting.py
Comment (schorrm) [inline pybaseball/team_batting.py:24]: same concern applies

File: pybaseball/lahman.py
Comment (schorrm) [inline pybaseball/lahman.py:8]: This should be untouched

File: pybaseball/pitching_leaders.py
Comment (schorrm) [inline pybaseball/pitching_leaders.py:23]: this could also be bool, if we're going to change the API around here (and that wouldn't break code that used 0/1)

File: pybaseball/pitching_leaders.py
Comment (schorrm) [inline pybaseball/pitching_leaders.py:33]: ??????

File: tests/pybaseball/analysis/projections/marcels/test_marcels_pitching.py
Comment (schorrm) [inline tests/pybaseball/analysis/projections/marcels/test_marcels_pitching.py:2]: This file should not be in the PR at all, what happened here?

File: tests/pybaseball/analysis/projections/marcels/test_marcels_batting.py
Comment (schorrm) [inline tests/pybaseball/analysis/projections/marcels/test_marcels_batting.py:2]: This file should not be in the PR at all, what happened here?

Comment (schorrm) [reply tests/pybaseball/analysis/projections/marcels/test_marcels_batting.py:2]: Also you just fudged the whitespace. Why did you add these to the commit in the first place?

File: pybaseball/team_pitching.py
Comment (schorrm) [inline pybaseball/team_pitching.py:24]: same

File: pybaseball/team_pitching.py
Comment (schorrm) [inline pybaseball/team_pitching.py:28]: why?

File: pybaseball/datasources/html_table_processor.py
Comment (schorrm) [inline pybaseball/datasources/html_table_processor.py:66]: Is there a reason we don't use the format `requests.get(url_endpoint, params=mydict)`?

**Review by schorrm — CHANGES_REQUESTED** [review]

File: pybaseball/datasources/fangraphs.py
Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py:76]: Because it can only ever be a 1 or 0. So it's really a boolean.

File: pybaseball/datasources/html_table_processor.py
Comment (TheCleric) [reply pybaseball/datasources/html_table_processor.py:68]: In theory yes, technically anything in the 2XX range is a success. As well anything in the 3XX range could be a redirect, though requests should follow those automatically to the real page.

So according to HTTP standards only status_codes >= 400 are errors. In actual practice we'll likely only be getting things in the 2XX range, 4XX range, or 5XX range. So we could make this `>299` just to catch weird redirects that requests missed, but otherwise the above is fairly standard.

File: pybaseball/batting_leaders.py
Comment (TheCleric) [reply pybaseball/batting_leaders.py:33]: Me not respecting the data type! 😆 

I'll fix.

Comment (TheCleric) [reply pybaseball/batting_leaders.py:33]: Existing code wants an int. This is not quite the correct data type for this field as it's either on (1) or off(0). So we're translating the existing data type to booleans to be more honest and specific about what that function accepts.

Also I need to compare to an int, not a string, which I will fix.

File: pybaseball/lahman.py
Comment (TheCleric) [reply pybaseball/lahman.py:8]: I'll undo. Sorry. Must have organized these imports.

File: pybaseball/pitching_leaders.py
Comment (TheCleric) [reply pybaseball/pitching_leaders.py:23]: It's true python would just use the truthiness of the 1 or 0, so the effect SHOULD be the same, for anyone who uses MyPy against their code to ensure data type consistency, that would raise an error.

I was erring on the side of not breaking existing function declarations for end user consistency.

File: tests/pybaseball/analysis/projections/marcels/test_marcels_pitching.py
Comment (TheCleric) [reply tests/pybaseball/analysis/projections/marcels/test_marcels_pitching.py:2]: Must have been checking something and some whitespace got added. Will fix.

File: tests/pybaseball/analysis/projections/marcels/test_marcels_batting.py
Comment (TheCleric) [reply tests/pybaseball/analysis/projections/marcels/test_marcels_batting.py:2]: Because you requested that I re-add the old testing, so these two files got caught up in the mix. Since I was just re-adding old files, I did not evaluate them individually.

File: pybaseball/datasources/html_table_processor.py
Comment (TheCleric) [reply pybaseball/datasources/html_table_processor.py:66]: Because I didn't know about that. Will add it in.

File: pybaseball/datasources/fangraphs.py
Comment (schorrm) [reply pybaseball/datasources/fangraphs.py:76]: okay, and to convert to int to pass it in requests just use `int(split_seasons)`, I guess

File: pybaseball/batting_leaders.py
Comment (schorrm) [reply pybaseball/batting_leaders.py:33]: That was addressing the ohter one -- why are separating batting_stats and fg_batting_data?

Comment (TheCleric) [reply pybaseball/batting_leaders.py]: batting_stats is the existing api. fg_batting_data is the new api. I did not want them named the same as I didn't want to cause confusion since they accept a different set of parameters.

If a user wants to continue using the old API it will continue to work. If they want to use the new one, it will be a conscious decision on their part.

File: pybaseball/batting_leaders.py
Comment (schorrm) [inline pybaseball/batting_leaders.py:33]: bool(ind)?

File: pybaseball/datasources/fangraphs.py
Comment (bdilday) [inline pybaseball/datasources/fangraphs.py:48]: I know I'm the one who removed it in my example PR, but could we put the doc string back in here?

Comment (bdilday) [inline pybaseball/datasources/fangraphs.py:48]: how does using a non-default value for `stats` work? I tried calling it with `fg_batting_data(2019, types=FanGraphsBattingStat.BABIP.value)` and got a 500 error. I guess you need the `c` to be inserted before the list of integer stats codes?

File: pybaseball/datasources/fangraphs.py
Comment (bdilday) [inline pybaseball/datasources/fangraphs.py:104]: If a user asks for non-default list of stats then `WAR` and / or `OPS` might not be valid columns. Could we handel that case here to avoid a possible `KeyError` when `sort_values` looks for a non-existing column?

File: pybaseball/datasources/fangraphs.py
Comment (bdilday) [reply pybaseball/datasources/fangraphs.py:48]: sorry, I meant non-default value for the argument, `types`

Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py:48]: CURSE YOU @bdilday.

Yeah I'll add it back in. I had intended to anyway.

Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py:48]: Yeah, it seems you either have to specify the c (which I think stands for common stats), or it's components (Name, Team, Season). I should at least document that, and possibly fix the call to insert that. What do you think?

File: pybaseball/datasources/fangraphs.py
Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py:104]: Very good point. I'll work something in.

File: pybaseball/datasources/fangraphs.py
Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py:48]: And now that I'm thinking further I need to revamp some of this code. If the request is for a fixed set of columns from the user, this shouldn't be a stat coming in, but rather a list of them. Thanks for helping me see a blind spot there.

Comment (bdilday) [reply pybaseball/datasources/fangraphs.py:48]: a suggestion - what if as a first step we don't give the user a choice, you just get all the stats. the user is free to filter the columns of the resulting data frame if they want. in a subsequent PR we could fix it, but it seems potentially complicated and I feel like deferring and merging this useful PR first is better than either blocking this PR or rushing a non-optimal solution

File: pybaseball/datasources/fangraphs.py
Comment (bdilday) [inline pybaseball/datasources/fangraphs.py:107]: I'm finding it a little confusing that `STATS` has a data type `FanGraphsStatTypes` and `TYPES` has a data type `FanGraphsStat`. Could we somehow make the names more consistent here? One thought is maybe the argument names for the method don't have to match the names in the fangraphs API? i.e. what the `types` query parameter is specifying is maybe better thought of table columns than "stats" (or "types" which is super ambiguous not tp mention conflict with the python protected keyword `type`)?

Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py:107]: Valid point. I'll rework.

File: pybaseball/datasources/fangraphs.py
Comment (schorrm) [reply pybaseball/datasources/fangraphs.py:48]: Yeah, I'll change my mind on this and side with @bdilday here

File: pybaseball/datasources/fangraphs.py
Comment (schorrm) [reply pybaseball/datasources/fangraphs.py:104]: Let's defer to as per @bdilday's other suggestion

File: pybaseball/datasources/fangraphs.py
Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py]: I didn't see these until after I had done some tweaks here. I now have it working consistently.

For example:

```python
import pybaseball
from pybaseball.enums.fangraphs import FanGraphsBattingStat

pybaseball.datasources.fangraphs.fg_batting_data(2019, types=[FanGraphsBattingStat.AB])
```

Now returns:

|Season | Name | Team | AB
|-- | -- | -- | --
|2019 | Christian Yelich | Brewers | 489
|2019 | Mike Trout | Angels | 470
|2019 | Alex Bregman | Astros | 554
|2019 | Nelson Cruz | Twins | 454
|2019 | Cody Bellinger | Dodgers | 558
|... | ... | ... | ...
|2019 | Khris Davis | Athletics | 481
|2019 | Yolmer Sanchez | White Sox | 496
|2019 | Brandon Crawford | Giants | 500
|2019 | Mallex Smith | Mariners | 510
|2019 | Orlando Arcia | Brewers | 494

Let me know if this works.

File: pybaseball/datasources/fangraphs.py
Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py]: This has already been tweaked, but we can discuss further.

File: pybaseball/datasources/fangraphs.py
Comment (bdilday) [reply pybaseball/datasources/fangraphs.py]: LGTM!

File: pybaseball/datasources/fangraphs.py
Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py]: Names have been reworked to be less confusing.

File: pybaseball/datasources/fangraphs.py
Comment (schorrm) [reply pybaseball/datasources/fangraphs.py]: Sorry, I _really_ hate to necromance this, but.
1. This is **really** good behavior to have.
2. We should allow:
```python
import pybaseball
pybaseball.fg_batting_data(2019, types=['AB'])
```

Comment (TheCleric) [reply pybaseball/datasources/fangraphs.py]: Already in there that way :)

stat_columns accepts either a str or List[str]

```python
import pybaseball

pybaseball.datasources.fangraphs.fg_batting_data(2019, stat_columns=['AB'])
```

|Season | Name | Team | AB
|-- | -- | -- | --
|2019 | Christian Yelich | Brewers | 489
|2019 | Mike Trout | Angels | 470
|2019 | Alex Bregman | Astros | 554
|2019 | Nelson Cruz | Twins | 454
|2019 | Cody Bellinger | Dodgers | 558
|... | ... | ... | ...
|2019 | Khris Davis | Athletics | 481
|2019 | Yolmer Sanchez | White Sox | 496
|2019 | Brandon Crawford | Giants | 500
|2019 | Mallex Smith | Mariners | 510
|2019 | Orlando Arcia | Brewers | 494

135 rows × 4 columns

**Pull request merged**
